aiohttp/connector.py

```diff
@@ -470,25 +477,34 @@ def _get(self, key):
         del self._conns[key]
         return None
 
+    def _release_key_waiter(self, key, waiters):
+        if not waiters:
+            return False
+
+        waiter = waiters.popleft()
```

Comment:
Umm, you have to persist till reaches a waiter not done or you run out of waiters, see the implementation of CPython [1].

[1] https://github.com/python/cpython/blob/master/Lib/asyncio/locks.py#L450

Reply:
are you saying the old implementation was wrong: https://github.com/aio-libs/aiohttp/pull/2937/files#diff-7f25afde79f309e2f8722c26cf1f10adL481 ?  I don't believe this is the case.  There are two ways a waiter can be removed:

1. An exception happened while waiting (in exception handler)
    1. a release was dispatched for said waiter (someone will see a release)
2. through this method

What you describe would create a double release for 1.i.  This is in fact the scenario you before alluded to

Reply:
Yes, the issue was already there, indeed I can see the following issues with the code that we have in master:

1) The iteration till reach a none canceled waiter has to be done through all of the items of a list, right now is only done on top of the head of each list.
2) Each time that we try to release a waiter we have to calculate if the limit and the number of concurrent connections allows us to make it. This is done only in one when the release_waiter is called explicitly but not when we had an exception trying to make the connection
3)The limit per host, TBH, i would say that is not well calculated.

So we have to fix them, but true that they were already there and would be nice if we decouple both things.

Reply:
ya I have a feeling this is the tip of the iceberg :)  I have a strong suspicious there's a leak in aiohttp or something aiohttp uses as right now we're leaking ~40MB / week in prod